roofit/roofitcore/inc/NumericalDerivatorMinuit2.h

```diff
@@ -0,0 +1,117 @@
+                            const std::vector<ROOT::Fit::ParameterSettings> &parameters);
+   std::vector<MinuitDerivatorElement>
+   Differentiate(const ROOT::Math::IBaseFunctionMultiDim *function, const double *x,
+                 const std::vector<ROOT::Fit::ParameterSettings> &parameters,
+                 const std::vector<MinuitDerivatorElement>& previous_gradient);
+
+   MinuitDerivatorElement
+   partial_derivative(const ROOT::Math::IBaseFunctionMultiDim *function, const double *x,
+                      const std::vector<ROOT::Fit::ParameterSettings> &parameters, unsigned int i_component, MinuitDerivatorElement previous);
+   MinuitDerivatorElement fast_partial_derivative(const ROOT::Math::IBaseFunctionMultiDim *function,
+                                const std::vector<ROOT::Fit::ParameterSettings> &parameters, unsigned int i_component, const MinuitDerivatorElement& previous);
+   MinuitDerivatorElement operator()(const ROOT::Math::IBaseFunctionMultiDim *function, const double *x,
+                                                 const std::vector<ROOT::Fit::ParameterSettings> &parameters,
+                                                 unsigned int i_component, const MinuitDerivatorElement& previous);
+
+   double GetFValue() const { return fVal; }
+   void SetStepTolerance(double value);
+   void SetGradTolerance(double value);
+   void SetNCycles(int value);
+
+   double Int2ext(const ROOT::Fit::ParameterSettings &parameter, double val) const;
+   double Ext2int(const ROOT::Fit::ParameterSettings &parameter, double val) const;
+   double DInt2Ext(const ROOT::Fit::ParameterSettings &parameter, double val) const;
+
+   void SetInitialGradient(const ROOT::Math::IBaseFunctionMultiDim *function,
+                           const std::vector<ROOT::Fit::ParameterSettings> &parameters,
+                           std::vector<MinuitDerivatorElement>& gradient);
+
+   void set_step_tolerance(double step_tolerance);
+   void set_grad_tolerance(double grad_tolerance);
+   void set_ncycles(unsigned int ncycles);
+   void set_error_level(double error_level);
+
+private:
+   double fStepTolerance = 0.5;
+   double fGradTolerance = 0.1;
+   unsigned int fNCycles = 2;
+   double Up = 1;
+   double fVal = 0;
+
+   std::vector<double> vx, vx_external;
+   double dfmin;
+   double vrysml;
+
+   // MODIFIED: Minuit2 determines machine precision in a slightly different way than
+   // std::numeric_limits<double>::epsilon()). We go with the Minuit2 one.
+   ROOT::Minuit2::MnMachinePrecision precision;
+
+   ROOT::Minuit2::SinParameterTransformation fDoubleLimTrafo;
+   ROOT::Minuit2::SqrtUpParameterTransformation fUpperLimTrafo;
+   ROOT::Minuit2::SqrtLowParameterTransformation fLowerLimTrafo;
+
+private:
+   bool _always_exactly_mimic_minuit2;
+
+public:
+   bool always_exactly_mimic_minuit2() const;
+   void set_always_exactly_mimic_minuit2(bool flag);
+
+private:
+   std::vector<double> vx_fVal_cache;
+};
```

Comment:
```suggestion
   std::vector<double> vx_fVal_cache;
   bool _always_exactly_mimic_minuit2;

public:
   bool always_exactly_mimic_minuit2() const;
   void set_always_exactly_mimic_minuit2(bool flag);
};
```
No padding bits between the `bool` and the `vector` and better grouping of members.